javascript/ql/src/Security/CWE-116/IncompleteSanitization.ql

```diff
@@ -106,8 +106,19 @@ where repl.getMethodName() = "replace" and
       (
        not old.(RegExpLiteral).isGlobal() and
        msg = "This replaces only the first occurrence of " + old + "." and
-       // only flag if this is likely to be a sanitizer
-       getAMatchedString(old) = metachar() and
+       // only flag if this is likely to be a sanitizer or URL encoder or decoder
+       exists (string m | m = getAMatchedString(old) |
+         // sanitizer
+         m = metachar()
+         or
+         exists (string urlEscapePattern | urlEscapePattern = "(%[0-9A-Fa-f]{2})+" |
+           // URL decoder
+           m.regexpMatch(urlEscapePattern)
+           or
+           // URL encoder
+           repl.getArgument(1).getStringValue().regexpMatch(urlEscapePattern)
```

Comment:
Not a real problem in this case, but it does bother me a bit that `m` isn't bound in the last case. It's not _that_ obvious that there is no cartesian product here.

Reply:
I can pull it out of the `exists` if you like and add `and exists(getAMatchedString(old))` instead (that bit is important to avoid false positives from cases where we cannot analyse `old`).

Reply:
Ah right, if the `exists` matters then let's keep it like this.